Client/WarFare/UILevelGuide.cpp

```diff
@@ -0,0 +1,385 @@
+
+		iCounter++;
+	}
+
+	CN3UIString* pTitles[3] = { m_pTextTitle0, m_pTextTitle1, m_pTextTitle2 };
+	CN3UIString* pGuides[3] = { m_pTextGuide0, m_pTextGuide1, m_pTextGuide2 };
+
+	//find max number of pages, ceil without cmath
+	int iMaxPage = m_iContentCount / 3 + (m_iContentCount % 3 != 0 ? 1 : 0);
+
+	if (m_iCurrentPage < 1) m_iCurrentPage = 1;
+
+	if (m_iCurrentPage > iMaxPage) m_iCurrentPage = iMaxPage;
+
+	int iStartIndex = (m_iCurrentPage - 1) * 3;
+
+	for (int i = 0; i < 3; i++)
+	{
+
+		if (iStartIndex >= iCounter) //last page 2 empty items
+		{
+			pTitles[i]->SetString("");
+			pGuides[i]->SetString("");
+			continue;
+		}
+
+		if (pTitles[i])
+			pTitles[i]->SetString(m_saQuestTitle[iStartIndex]);
+
+		if (pGuides[i])
+			pGuides[i]->SetString(m_saQuestText[iStartIndex]);
+
+
+
+		iStartIndex++;
+	}
+
+	if (m_pTextPage != nullptr)
+	{
+		std::string strCurrentPage = std::to_string(m_iCurrentPage);
+		m_pTextPage->SetString(strCurrentPage);
+	}
+
+
+
+}
+
+bool CUILevelGuide::IsNumber(const std::string& s)
+{
+	// empty string
+	if (s.empty()) return false;
+
+	// check is only digits
+	return std::all_of(s.begin(), s.end(), ::isdigit);
+}
+
+void CUILevelGuide::SetTopLine(CN3UIScrollBar* pScroll, CN3UIString* pTextGuide)
+{
+	if (pTextGuide == nullptr || pScroll == nullptr) return;
+
+	//total number of lines of text
+	const int iTotalLineCount = pTextGuide->GetLineCount();
+
+	//max number of lines visible in text area
+	const int iVisibleLineCount = 4;
+
+	//return if text is shorter than or equal to 4 lines
+	if (iTotalLineCount <= iVisibleLineCount)
+		return;
+
+	//topline,size_t
+	int iTopLine = static_cast<int>(pScroll->GetCurrentPos());
+
+	// limit check for the line which displayed first, topline
+	if (iTopLine < 0)
+		iTopLine = 0;
+	else if (iTopLine > iTotalLineCount - iVisibleLineCount)
+		iTopLine = iTotalLineCount - iVisibleLineCount;
+
+
+	pTextGuide->SetStartLine(iTopLine);
+
+}
+
+void CUILevelGuide::SetVisible(bool bVisible)
+{
+	CN3UIBase::SetVisible(bVisible);
+	if (bVisible)
+	{
+		//display initial user info with respect to user level
+		m_iCurrentPage = 1;
+		LoadContent();
+		CGameProcedure::s_pUIMgr->SetVisibleFocusedUI(this);
+	}
+	else
+	{
+		CGameProcedure::s_pUIMgr->ReFocusUI();//this_ui
+		if (m_pEditLevel && m_pEditLevel->HaveFocus())
+			m_pEditLevel->KillFocus();
+	}
+}
+
+bool CUILevelGuide::ReceiveMessage(CN3UIBase* pSender, uint32_t dwMsg)
+{
+	if (dwMsg == UIMSG_BUTTON_CLICK)
+	{
+		if (pSender == (CN3UIBase*) m_pButtonCancel)
+		{
+			this->SetVisible(false);
+			return true;
+		}
+		else if (pSender == (CN3UIBase*) m_pButtonUp)
+		{
+			m_iCurrentPage++;
+			LoadContent();
+			return true;
+		}
+		else if (pSender == (CN3UIBase*) m_pButtonDown)
+		{
+			m_iCurrentPage--;
+			LoadContent();
+			return true;
+		}
+		else if (pSender == (CN3UIBase*) m_pButtonCheck) //search button
+		{
+
+			std::string strSearchLevel;
```

Comment:
Realistically I would restructure this as explained previously (drop IsNumber(), filter numerical input only in the control instead, use atoi() and not bother dealing with exceptions), but that aside, this should be a const reference (it would need to be reshuffled slightly).